src/Query.ts

```diff
@@ -294,222 +297,203 @@ export class Query {
         ? { searchProperty: { $elemMatch: { $in: searchProperty } } }
         : { includeInGlobalSearch: true };
-    const compound: { should: any[]; must?: any[]; filter?: any[]; minimumShouldMatch: number } = {
-      should: parts,
-      minimumShouldMatch: 1,
-    };
 
-    // if any selected facets. add to filter part of compound operator
-    compound.filter = getFiltersFromSelections(selectedFacets);
+    const compound = this.getCompound();
 
-    const facets = getFacets(selectedFacets, taxonomy);
-
-    const agg: any = [
+    const agg = [
       {
         $search: {
-          facet: {
-            operator: {
-              compound: compound,
-            },
-            facets: facets,
+          compound,
+          tracking: {
+            searchTerms: this.rawQuery,
           },
         },
       },
       { $match: filter },
     ];
-    agg.push({
-      $facet: {
-        docs: [
-          {
-            $project: {
-              _id: 0,
-              title: 1,
-              preview: 1,
-              url: 1,
-              searchProperty: 1,
-            },
-          },
-          {
-            $limit: 50,
-          },
-        ],
-        meta: [{ $replaceWith: '$$SEARCH_META' }, { $limit: 1 }],
-      },
-    });
-    agg.push({
-      $unwind: {
-        path: '$meta',
-      },
-    });
     console.log('Executing ' + JSON.stringify(agg));
     return agg;
   }
 }
 
-/**
- * gets filters required for compound operator
- * https://www.mongodb.com/docs/atlas/atlas-search/compound/#filter-examples
- */
-const getFiltersFromSelections = (selectedFacets: string[]) => {
-  const filters = [];
-  for (const selectedFacetKey of selectedFacets) {
-    // for each selected facet key ie. (languages←python→versions←3.7)
-    // split for the last ← key and left is key, right is value
-    const splitLet = [...selectedFacetKey];
-    let idx = splitLet.length - 1;
-    let targetIdx;
-    while (!targetIdx && idx > -1) {
-      if (splitLet[idx] === '←') {
-        targetIdx = idx;
-        break;
-      }
-      idx--;
+export const extractFacetFilters = (searchParams: URL['searchParams']): Filter<Document> => {
+  // query should be in form of:
+  // q=test&facets.target_platforms>manual>versions=v6.0&facets.target_platforms=atlas
+  // where each query param starting with 'facets.' denotes a filter and possible expansion
+  // {
+  //   "facets.target_platforms": ["manual", "atlas"],
+  //   "facets.target_platforms>manual>versions": ["v6.0"]
+  // }
+  const filter: Filter<Document> = {};
+  for (const [key, value] of searchParams) {
+    if (!key.startsWith('facets.')) {
+      continue;
     }
+    const facetNames = key.replace('facets.', '').split('>');
+    for (let facetIdx = 0; facetIdx < facetNames.length; facetIdx++) {
+      // hierarchy facets denoted by >
+      // each facet node requires a facet property, at every even level
+      if (facetIdx % 2 === 1) continue;
```

Comment:
(Nit) I liked your logic in the other `for` loop where we just increment by 2. Is that something we can do here as well? (but keep the comment, since it's good context)